README.md
# PostCSS srcset plugin [![npm version](https://img.shields.io/npm/v/postcss-srcset.svg?style=flat)](https://www.npmjs.com/package/postcss-srcset)

<img align="right" width="135" height="95"
     title="Philosopher's stone, logo of PostCSS"
     src="http://postcss.github.io/postcss/logo-leftp.svg">

Assists in auto-generating background images scaled for different viewport
widths.

Install with npm: `npm install --save-dev postcss-srcset`

## What it does

The postcss-srcset plugin provides a CSS function `srcset` that takes an image
path and list of viewport widths. For each width, the plugin will create a
media query that applies to that viewport width (and below) and contains
a duplicate of the declaration it's used in, with the width appended to the URL
in a query string.

The plugin doesn't actually scale the images, but the package includes a webpack
loader that you can use to do the job.

## Examples

```javascript
// Code example
const postcss = require('postcss')

postcss([
  require('postcss-srcset')(/* optional options */)
])
```

```css
/* Input example */
html {
  background: srcset('./myimage.png', 300 500 1000) cover;
  color: white;
}
```

```css
/* Output example */
html {
  background: url('./myimage.png') cover;
  color: white;
}

@media (max-width: 1000px) {
  html {
    background: url('./myimage.png?size=1000') cover;
  }
}

@media (max-width: 500px) {
  html {
    background: url('./myimage.png?size=500') cover;
  }
}

@media (max-width: 300px) {
  html {
    background: url('./myimage.png?size=300') cover;
  }
}
```

For a more comprehensive example, see the [demo](/demo).

### Options

You can pass an options object as the first argument to the plugin. Here is
an explanation of the available configuration arguments and their default
values:

```javascript
const options = {
  // The name of the function to make available in the css, in case you take
  // offense to my default naming of it.
  function: 'srcset',
  // The function that transforms the image's URL for a given media query.
  // You can return a promise from here if you want to do something async.
  transformUrl: (url, size) => `${url}?size=${size}`
}

postcss([
  require('postcss-srcset')(options)
])
```

### Demo

To run the demo, clone this repo and run `npm i && npm start`.

![Demonstration of the output](demo.gif)

## Use With Build Tools

This plugin comes with a loader for webpack, but you can use any build tool
you like.

### Webpack

An example of using the built-in webpack loader:

```javascript
const srcsetPlugin = require('postcss-srcset')

module.exports = {
  module: {
    rules: [
      {
        test: /\.css$/,
        use: [
          'style-loader',
          'css-loader',
          {
            loader: 'postcss-loader',
            options: {
              plugins: [srcsetPlugin()]
            }
          }
        ]
      },
      {
        test: /\.(jpe?g|png)$/,
        resourceQuery: /[?&]size(=|&|\[|$)/,
        use: [
          'url-loader',
          srcsetPlugin.loader
        ]
      }
    ]
  }
}
```

### Other Build Tools

If you want to use another build tool you may have to write it yourself.
Unless you're doing something very simple, in which case you can do the image
resizing just in the plugin's `transformUrl` argument.

Either way, if you're happy using the same image resize function that the
webpack loader uses, it's available for you to `require`:

```javascript
const resizeImage = require('postcss-loader/resize-image');
const fs = require('fs')

var image = fs.readFileSync('myImage.png')
// Arguments: Image buffer, image width
// Returns a promise that resolves to a buffer with the resized image
resizeImage(image, 400).then(
  resized => fs.writeFileSync('myImage-400.png', resized),
  error => console.error('Error in resizeImage:', error)
)
```

## ACK

This file is inspired by (and partly copied from) the [srcset-loader][1];
check that out if you want to use `srcset` in img tags.

[1]: https://github.com/timse/srcset-loader/


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.1 - 2017-05-11
### Added
 - A basic demo showing usage of the plugin & loader
 - Moved `peerDependencies` to `dependencies`

## 1.0.0 - 2017-05-11
### Added
 - Basic implementation of plugin and webpack loader


tests/loader.test.js
const mockSharp = jest.fn();
const mockResize = jest.fn();

mockSharp.mockReturnValue({
  resize: mockResize
});
mockResize.mockReturnValue({
  toBuffer: () => Promise.resolve("output")
});

jest.mock("image-size", () => () => 10000);
jest.mock("sharp", () => mockSharp);

const loader = require("../resize-loader");

test("resize-loader works", () =>
  new Promise((res, rej) => {
    loader.call(
      {
        query: {},
        resourceQuery: "?size=100",
        async: () => (err, buffer) => {
          if (err) rej(err);
          else res(buffer);
        }
      },
      "input"
    );
  }).then(buffer => {
    expect(mockSharp).toBeCalledWith("input");
    expect(mockResize).toBeCalledWith(100);
    expect(buffer).toEqual("output");
  }));


tests/plugin.test.js
const postcss = require('postcss');
const srcset = require('..');

const testInput = `html {
  background: srcset('./images/background.jpg', 300 600 900) cover;
  color: white;
}
`;

const testOutput = require('fs').readFileSync(
  `${__dirname}/test-output.css`,
  'utf-8'
);

it('Converts srcset into multiple media queries', () =>
  postcss([srcset()])
    .process(testInput)
    .then(result => expect(result.css).toEqual(testOutput)));


resize-loader.js
// This is more or less lifted from srcset-loader
// https://github.com/timse/srcset-loader/blob/master/src/resize-loader.js
var loaderUtils = require('loader-utils');
var resizeImage = require('./resize-image');

function parseQuery(query) {
  if (!query) {
    return {};
  }

  if (typeof query === 'string') {
    return loaderUtils.parseQuery(query);
  }

  return query;
}

module.exports = function resizeLoader(content) {
  if (this.cacheable) {
    this.cacheable();
  }

  const options = loaderUtils.getOptions(this);

  const queryParam = (options && options.queryParam) || 'size';

  const callback = this.async();

  const query = parseQuery(this.resourceQuery);

  const size = parseInt(query[queryParam], 10);

  resizeImage(content, size).then(
    buffer => callback(null, buffer),
    err => callback(err)
  );
};

module.exports.raw = true;


index.js
const postcss = require('postcss');
const balancedMatch = require('balanced-match');
const PLUGIN_NAME = 'postcss-srcset';

/**
 * Srcset transformer plugin
 *
 * Use in this form:
 * .cls {
 *   background-image: srcset('./path/to/my/image.png', 300 500 1200);
 * }
 *
 * The first argument is the path to your image, the second argument is a
 * space-separated list of sizes to scale the image to.
 *
 * The function will be converted to a collection of media queries that provide
 * a different URL at each size.
 *
 * OPTIONS:
 * function - the name of the function to transform. Defaults to `srcset`
 * transformUrl - the function to transform the image url for the given size
*                 defaults to (url, size) => `${url}?cssSize=${size}`
 *                can return a promise
 */
module.exports = postcss.plugin(
  PLUGIN_NAME,
  (options = {}) =>
    function(css) {
      // In case someone wants to use an async transform function
      var promises = [];

      // Set default options
      var funcName = options.function || 'srcset';
      var transformUrl =
        options.transformUrl || ((url, size) => `${url}?size=${size}`);

      // Look for declerations that use srcset
      css.walkDecls((decl) => {
        if (decl.value.indexOf(funcName) === -1) {
          return;
        }

        // Balanced match in case there are any functions-in-functions
        // (shouldn't be but doesn't hurt to check)
        let match = balancedMatch(`${funcName}(`, ')', decl.value);

        // Can't do layers yet, need to figure out how different size sets
        // combine into each media query
        if (match.post.indexOf(`${funcName}(`) > -1) {
          throw decl.error(
            `${PLUGIN_NAME} does not currently support more than one ` +
              `${funcName} call per declaration`
          );
        }

        let [url, sizes] = postcss.list.comma(match.body);

        // Strip string delemiters from start/end
        url = url.replace(/^['"]|['"]$/g, '');

        // Sizes to integer and in reverse order so the smallest sizes
        // goes last and thus gets the highest specificity
        sizes = postcss.list
          .space(sizes)
          .map(n => parseInt(n, 10))
          .sort((a, b) => b - a);

        // Url back to value (including any values that became before or after)
        let unparse = url => `${match.pre}url('${url}')${match.post}`;

        promises.push(
          ...sizes.map(size =>
            Promise.resolve(transformUrl(url, size)).then(url =>
              // Construct the new rules for each size
              constructRule(decl, unparse(url), size)
            )
          )
        );

        // The default value should be the image at its largest (original) size
        decl.value = unparse(url);
      });

      return Promise.all(promises);
    }
);

/**
 * Make a media query at-rule for a given size
 */
function makeQuery(size) {
  return postcss.atRule({
    name: 'media',
    params: `(max-width: ${size}px)`
  });
}

/**
 * Clone a rule and remove all of its declarations
 */
function cloneRule(decl) {
  return decl.parent.clone({
    nodes: []
  });
}

/**
 * clone a delcaration and assign it a new value
 */
function cloneDecl(decl, value) {
  var d = decl.clone({});

  d.value = value;

  return d;
}

/**
 * Construct a new media query with rule and declaration
 * and attatch it to the root
 */
function constructRule(decl, value, size) {
  var query = makeQuery(size);
  var rule = cloneRule(decl);
  var newDecl = cloneDecl(decl, value);
  query.source = decl.source;

  rule.append(newDecl);
  query.append(rule);
  decl.root().append(query);
}

module.exports.default = module.exports;
module.exports.loader = require.resolve('./resize-loader');
